components/scream/src/share/grid/remap/coarsening_remapper.cpp

```diff
@@ -0,0 +1,965 @@
+                             KOKKOS_LAMBDA(const MemberType& team){
+          const int i = team.league_rank();
+          const int lid = lids_pids(i,0);
+          const int pid = lids_pids(i,1);
+          const int lidpos = i - pid_lid_start(pid);
+          const int offset = f_pid_offsets(pid);
+
+          Kokkos::parallel_for(Kokkos::TeamThreadRange(team,ndims*nlevs),
+                               [&](const int idx) {
+            const int idim = idx / nlevs;
+            const int ilev = idx % nlevs;
+            buf(offset + lidpos*ndims*nlevs + idim*nlevs + ilev) = v(lid,idim,ilev);
+          });
+        });
+      } break;
+
+      default:
+        EKAT_ERROR_MSG ("Unexpected field rank in CoarseningRemapper::pack.\n"
+            "  - MPI rank  : " + std::to_string(m_comm.rank()) + "\n"
+            "  - field rank: " + std::to_string(fl.rank()) + "\n");
+    }
+  }
+
+  // If MPI does not use dev pointers, we need to deep copy from dev to host
+  if (not MpiOnDev) {
+    Kokkos::deep_copy (m_mpi_send_buffer,m_send_buffer);
+  }
+
+  if (not m_send_req.empty()) {
+    int ierr = MPI_Startall(m_send_req.size(),m_send_req_ptr);
+    EKAT_REQUIRE_MSG (ierr==MPI_SUCCESS,
+        "Error! Something whent wrong while starting persistent send requests.\n"
+        "  - send rank: " + std::to_string(m_comm.rank()) + "\n");
+  }
+}
+
+void CoarseningRemapper::recv_and_unpack () const
+{
+  if (not m_recv_req.empty()) {
+    int ierr = MPI_Waitall(m_recv_req.size(),m_recv_req_ptr, MPI_STATUSES_IGNORE);
+    EKAT_REQUIRE_MSG (ierr==MPI_SUCCESS,
+        "Error! Something whent wrong while waiting on persistent recv requests.\n"
+        "  - recv rank: " + std::to_string(m_comm.rank()) + "\n");
+  }
+  // If MPI does not use dev pointers, we need to deep copy from host to dev
+  if (not MpiOnDev) {
+    Kokkos::deep_copy (m_recv_buffer,m_mpi_recv_buffer);
+  }
+
+  using RangePolicy = typename KT::RangePolicy;
+  using MemberType  = typename KT::MemberType;
+  using ESU         = ekat::ExeSpaceUtils<typename KT::ExeSpace>;
+
+  const auto pid_lid_start = m_recv_pid_lids_start;
+  const auto lids_pids = m_recv_lids_pids;
+  const auto buf = m_recv_buffer;
+  for (int ifield=0; ifield<m_num_fields; ++ifield) {
+    const auto& f  = m_tgt_fields[ifield];
+    const auto& fl = f.get_header().get_identifier().get_layout();
+    const auto lt = get_layout_type(fl.tags());
+    const auto f_pid_offsets = ekat::subview(m_recv_f_pid_offsets,ifield);
+
+    f.deep_copy(0);
+    switch (lt) {
+      case LayoutType::Scalar2D:
+      {
+        auto v = f.get_view<Real*>();
+        Kokkos::parallel_for(RangePolicy(0,m_total_num_recv_gids),
+                             KOKKOS_LAMBDA(const int& i){
+          const int lid = lids_pids(i,0);
+          const int pid = lids_pids(i,1);
+          const int lidpos = i - pid_lid_start(pid);
+          const int offset = f_pid_offsets(pid);
+
+          v(lid) += buf (offset + lidpos);
```

Comment:
In general, multiple remotes will update the owner's `v(lid)`. Thus, isn't this loop (and the subsequent similar ones) a write race on `v(lid)`?

Reply:
👆

Reply:
Ah, good catch! It is indeed a race condition. I need to rearrange the unpacking, and use a team ||for.